File: 251001/Karpekov/lab1/TaskRest/src/main/java/lab1/services/TagService.java
package lab1.services;

import lab1.datalayer.DataLayer;
import lab1.datalayer.FieldChecker;
import lab1.datalayer.InMemoryDataLayer;
import lab1.models.Tag;
import org.springframework.stereotype.Component;

import java.util.ArrayList;

@Component
public class TagService {

    private final DataLayer dataLayer;
    public TagService(){
        dataLayer = new InMemoryDataLayer();
    }

    public ArrayList<Tag> GetTags(){
       return dataLayer.GetTags();
    }

    public Tag GetTagById(long Id) throws IllegalArgumentException{
        Tag res = dataLayer.GetTag(Id);
        return res;
    }

    public boolean CreateTag(Tag tag){
        if (FieldChecker.CheckTagFields(tag)){
            dataLayer.CreateTag(tag);
            return tag.getId() != -1;
        }else
            return false;
    }

    public Tag DeleteTag(long Id){
        return dataLayer.DeleteTag(Id);
    }

    public Tag UpdateTag(Tag tag){
        if (FieldChecker.CheckTagFields(tag))
            return dataLayer.UpdateTag(tag);
        else
            return null;
    }

    public DataLayer getDataLayer(){
        return dataLayer;
    }

}

File: 251002/Troitsky/Lab1-5/Lab1.Application/Services/TagService.cs
using Lab1.Core.Abstractions;
using Lab1.Core.Models;
using Lab1.Core.Contracts;
using System.Numerics;
using FluentValidation;
using Lab1.Application.Exceptions;
namespace Lab1.Application.Services
{
    public class TagService : ITagService
    {
        private ITagRepository _repository;
        private IValidator<Tag> _validator;
        public TagService(ITagRepository repository, IValidator<Tag> validator) => (_repository, _validator) = (repository, validator);
        public TagResponseTo CreateTag(Tag tag)
        {
            var res = _validator.Validate(tag);
            if (!res.IsValid)
                throw new IncorrectDataException("invalid input");
            try
            {
                return _repository.Create(tag);
            }
            catch
            {
                throw new IncorrectDatabaseException("DB error");
            }
        }
        public bool DeleteTag(ulong id)
        {
            try
            {
                return _repository.Delete(id);
            }
            catch
            {
                throw new IncorrectDatabaseException("DB error");
            }
        }
        public TagResponseTo? GetTag(ulong id)
        {
            try
            {
                return _repository.Get(id);
            }
            catch
            {
                throw new IncorrectDatabaseException("DB error");
            }
        }
        public List<TagResponseTo> GetAllTags()
        {
            try
            {
                return _repository.GetAll();
            }
            catch
            {
                throw new IncorrectDatabaseException("DB error");
            }
        }
        public TagResponseTo? UpdateTag(Tag tag, ulong id)
        {
            var res = _validator.Validate(tag);
            if (!res.IsValid)
                throw new IncorrectDataException("invalid input");
            try
            {
                return _repository.Update(tag, id);
            }
            catch
            {
                throw new IncorrectDatabaseException("DB error");
            }
        }
    }
}
